test/Sema/exhaustive_switch.swift

```diff
@@ -1261,13 +1264,15 @@ enum SR11212Tests {
     switch u {
     case .upair(let u_): return u_
     // expected-warning@-1 {{cannot match several associated values at once, implicitly tupling the associated values and trying to match that instead}}
+    // expected-note@-2 {{'upair' has 2 associated values but there is only 1 pattern here}}
```

Comment:
> This test case is a bit different because there's nothing here that actually suggests the user is thinking of the associated value as a tuple.

The cases we've seen in the wild before (e.g. in [SwiftPM](https://github.com/apple/swift/pull/26357#issuecomment-515254117)) seems to imply that the users are using the associated values as a tuple. One of our [test cases](https://github.com/apple/swift/pull/28076/commits/de1cf4b056ec4343daf2312068b37c2f8e09e8c5#diff-377347b625fb8d676bbc6bbc7dbc1d3a) had this too.

> The developer may have made a mistake, like forgetting the number of associated values or changing the declaration without updating this use site.

I agree that this is possible, in which case, it would be wrong for us to emit this diagnostic.

Reply:
> The cases we've seen in the wild before (e.g. in SwiftPM) seems to imply that the users are using the associated values as a tuple. One of our test cases had this too.

It's certainly possible to use this accidental feature intentionally. My point is that, if you write `case let .foo((x, y))`, we can pretty much assume that you were intending to treat the associated value as a tuple. If you write `case let .foo(x)`, we need to emit a diagnostic that allows for the possibility that you forgot about `y`.

-----

I suppose we could use a variant of @xedin's wording here:

```
enum case 'upair' expects 2 separate arguments; matching them as a single tuple is deprecated
```

But that means we need to discuss:

> "deprecated" -- this might make a user think "wait, don't we have stability guarantees, how can they deprecate stuff like that?" or "if this was working just fine before, why'd they deprecate it without any good reason? Swift devs bad"

And the answer is: We added a diagnostic because we discovered that this behavior could be confusing, for instance, when you add a second associated value and all of your existing switches suddenly start binding tuples. We made it a warning instead of an error because we guarantee source stability. Deprecation *is* source stability—it's how we fix mistakes without breaking our promise that your code will keep building in future compilers.

That answer won't make everyone happy, but there is no answer that will make everyone happy—not even removing this warning and blessing the implicit conversion as an "official" feature. Sometimes we just have to let some people be upset about a decision.